### DISCUSSION TITLE
Pest 3 Working Group

### DISCUSSION THREAD

**Discussion by MaxBondABE** [pest-parser/pest#640] [General]
This issue is to coordinate the formation of a working group for releasing a new version of Pest.

It is clear from the issues in this tracker that there is an appetite for a new version of Pest (#491 #485 ) . @dragostis and @CAD97 have also expressed that they need additional manpower to achieve this (https://github.com/pest-parser/pest/issues/491#issuecomment-780284980). After some discussion on the Glitter, we decided to create this issue and recruit any interested parties to form a new release of Pest.

**Comment by MaxBondABE** [comment]
I would propose the following initial set of priorities:

* Test & release the current version in this repository, to address the security concerns discussed here https://github.com/pest-parser/pest/issues/491#issuecomment-788182509
* Triage the backlog of issues & PRs
* Work through any critical issues & PRs
* Update the documentation to reflect these changes
* Finish the Pest book

I'm sure there's more to do after that, but these are the things I see as most critical. What do other people think?

**Comment by CAD97** [comment]
Just FYI: I'm not on the gitter anymore (other obligations, etc), but I'm still actively monitoring the repo and can review/merge small stuff given a bit of time.

I'm in favor of sharing the load, though. If only I had more time to give 😅

**Comment by MarinPostma** [comment]
Same as @CAD97 here. We use pest at MeiliSearch, and are happy to help maintain/develop the crate.

**Comment by dragostis** [comment]
I've added @MaxBondABE and @MarinPostma to the organization. A good start would be to fix the CI (even without the fuzzing), merge any relevant PRs and make a new pest release.

After this, it would make sense to talk about the future of the project somewhere. Since Gitter seems less used, maybe it makes sense to migrate somewhere more accessible?

**Comment by MarinPostma** [comment]
Thank you @dragostis !

Most rust communities I'm in use discord, we could do the same?

I can take a look at the CI

**Comment by dragostis** [comment]
Discord sounds good. I'll try to set it up later today.

**Comment by dragostis** [comment]
I've made a server here: https://discord.gg/bhSeAw3QPD Feel free to suggest any changes to it.

**Comment by 01mf02** [comment]
I would be also happy to help with the development of Pest, in particular working through existing PRs.
And I also have some ideas for completing the Pest book, as I have implemented [something similar to its proposed "final project"](https://github.com/01mf02/jaq).

**Comment by rbtcollins** [comment]
If I may make a suggestion as a small pest user (and willing to do the odd patch); don't block releases on future work. Unless the content in the repository right now is strictly worse than the previous release, just doing a release is an immediate benefit.

**Comment by bryantbiggs** [comment]
I left a message on the discord server as well, but how can I help? specifically start with https://github.com/pest-parser/pest/issues/485 and see what traction we can get rolling again

**Comment by tomtau** [comment]
The new release is out: https://github.com/pest-parser/pest/discussions/669 FYI

**Comment by Tartasprint** [comment]
I may be wrong but it's seems to me that there are many issues for the [v3.0 milestone](/pest-parser/pest/milestone/6) that have been pending for a long time. I do understand that it might be because the maintainers do not have enough time. But it seems that there are also two contradictory wishes: one is to make new changes such as #815, another is not to break things. But as noted in [this comment](https://github.com/pest-parser/pest/pull/815#issuecomment-1452771364)  the  [repository for the 3.0 version](/pest-parser/pest3) is dead.

I would like to contribute to this project, by helping to address some of the issues for the [v3.0 milestone](/pest-parser/pest/milestone/6), but I don't know where those changes should be put. Also if you have any recommendation to where efforts are needed, I'd be glad to hear them.

I realize this comment might sound like a critic, but it is not my wish, I think this project is awesome :D

**Reply by tomtau** [reply]
I may have "overclassified" some of the issues -- some of them may be possible, perhaps in some reduced forms, to be non-breaking even in the existing codebase. Having said that, API and grammar breakages aren't the only consideration; it's also the fact that the current surface grammar and its rewriting rules are quite complex, and this complexity may unnecessarily grow through interactions with those changes.

Anyway, as for potential contributions, my suggestion for starting them is the following:

- Pick issues you find interesting and comment on them; we can discuss there whether it'd be feasible within the current codebase.
- If it's not feasible, it'll be best done in a clean-slate repo first, i.e. reviving the v3 effort -- so you can either continue in that pest3 repo or jettison its code and start from scratch there. As I mentioned on Discord etc., I won't likely have time to contribute code to pest3 in the upcoming months, but am happy to review potential PRs there.

As for a recommendation where efforts are needed, I recall, for example, this issue had a potential idea for a "backwards-compatible" change to the current grammar: https://github.com/pest-parser/pest/issues/271#issuecomment-1224662176 (I don't know if @Sytten had time to work on it.)

**Reply by Sytten** [reply]
I didn't get to it sadly, got pulled into my startup after I found a work around. But it would be a valuable change for sure. The current limitations with whitespaces are annoying and the work around was not that great.
